core/src/repair_service.rs

```diff
@@ -179,45 +182,41 @@ impl RepairService {
         blocktree: &Blocktree,
         max_repairs: usize,
         repair_info: &mut RepairInfo,
-        repair_range: &RepairSlotRange,
     ) -> Result<(Vec<RepairType>)> {
         // Slot height and blob indexes for blobs we want to repair
         let mut repairs: Vec<RepairType> = vec![];
-        let mut current_slot = Some(repair_range.start);
-        while repairs.len() < max_repairs && current_slot.is_some() {
-            if current_slot.unwrap() > repair_range.end {
-                break;
-            }
 
-            if current_slot.unwrap() > repair_info.max_slot {
-                repair_info.repair_tries = 0;
-                repair_info.max_slot = current_slot.unwrap();
-            }
-
-            if let Some(slot) = blocktree.meta(current_slot.unwrap())? {
-                let new_repairs = Self::process_slot(
-                    blocktree,
-                    current_slot.unwrap(),
-                    &slot,
-                    max_repairs - repairs.len(),
-                );
-                repairs.extend(new_repairs);
+        // Iterate through the possible forks in order (they are weighted by lockout)
+        for (slot, _) in &repair_info.forks {
```

Comment:
@aeyakovenko, @rob-solana, this loop will try to repair any slots reachable by traversing the children of the slots in repair_info.forks. 

Should we try to preemptively ask for the next slots? For instance if this fork is present in blocktree: 1->3->5, and we finish repairing 1, 3, 5, should we ask the network for children of 5? This is to cover the case where we miss an entire slot/are trying to catch up with the rest of the network.

Reply:
no pre-emptive repair for now

Reply:
@rob-solana, I think without pre-emptive repairs some tests that test validator catchup on startup may fail. For instance we have some tests that write a few slots N to a ledger, and then wait for validators to catch up. Then when the validator starts up they see slots N + 1 onwards, but won't know how to chain those to their set of active forks (they start with only slot 0 as the active fork)

Reply:
@sagar-solana

Reply:
I don't mind leaving the dumb preemptive logic in to begin with (try to repair the next slot).
Although the more efficient way would be to request the highest blob for each fork a node has observed. The downside being that it is going to be serving 1 blob for ever fork its seen and that might end up being a lot depending on how far behind the repair request is.

Reply:
cool!